google/cloud/sql/connector/InstanceConnectionManager.py

```diff
@@ -439,19 +425,136 @@ def immediate_future(object: Any) -> concurrent.futures.Future:
+        :param username: A string representing the username use to connect to the
+            Cloud SQL instance.
 
-        :rtype: OpenSSl.SSL.Connection
-        :returns: An OpenSSL connection to the primary IP of the database.
+        :returns: A DB-API connection to the primary IP of the database.
         """
+        logger.debug("Entered connect method")
+        if username is None:
+            raise ValueError("No username provided!")
+
         with self._lock:
             instance_data: InstanceMetadata = self._current.result()
-        ctx = instance_data.ssl_context
-        ip_addr = instance_data.ip_addresses["PRIMARY"]
-        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-        ssl_connection = OpenSSL.SSL.Connection(ctx, s)
-        ssl_connection.connect((ip_addr, 3307))
-        ssl_connection.do_handshake()
-
-        return ssl_connection
+
+        try:
+            connector = {
+                "pymysql": self._connect_with_pymysql,
+                "pg8000": self._connect_with_pg8000,
+            }[driver]
+        except KeyError:
+            raise KeyError("Driver {} is not supported.".format(driver))
+
+        return connector(
+            instance_data.filepaths["ca"],
+            instance_data.filepaths["cert"],
+            instance_data.filepaths["key"],
+            instance_data.ip_addresses,
+            username,
+            **kwargs
+        )
+
+    def _connect_with_pymysql(
+        self,
+        ca_filepath: str,
+        cert_filepath: str,
+        key_filepath: str,
+        ip_addresses: Dict[str, str],
+        username: str,
+        **kwargs
+    ):
+        """Helper function to create a pymysql DB-API connection object.
+
+        :type ca_filepath: str
+        :param ca_filepath: A string representing the path to the server's
+            certificate authority.
+
+        :type cert_filepath: str
+        :param cert_filepath: A string representing the path to the ephemeral
+            certificate.
+
+        :type key_filepath: str
+        :param key_filepath: A string representing the path to the private key file.
+
+        :type ip_addresses: Dict[str, str]
+        :param ip_addresses: A Dictionary containing the different IP addresses
+            of the Cloud SQL instance.
+
+        :type username: str
+        :param username: A string representing the username use to connect to the
+            Cloud SQL instance.
+
+        :rtype: pymysql.Connection
+        :returns: A PyMySQL Connection object for the Cloud SQL instance.
+        """
+        import pymysql
+
+        ssl_dict = {
+            "ssl": {"ca": ca_filepath, "cert": cert_filepath, "key": key_filepath}
+        }
+
+        logger.debug(
+            "Temporary files: {}, {} and {}".format(
+                ca_filepath, cert_filepath, key_filepath
+            )
+        )
+
+        return pymysql.connect(
+            host=ip_addresses["PRIMARY"], user=username, ssl=ssl_dict, **kwargs
```

Comment:
What happens if `host` is passed in kwargs? Does it raise an error?

Reply:
It should, since we want to override host no matter what